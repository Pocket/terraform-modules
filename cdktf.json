{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
