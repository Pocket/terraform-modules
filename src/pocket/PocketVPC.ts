import { Fn, Resource } from 'cdktf';
import { AwsProvider, datasources, kms, ssm, vpc } from '@cdktf/provider-aws';
import { Construct } from 'constructs';

export class PocketVPC extends Resource {
  public readonly vpc: vpc.DataAwsVpc;

  public readonly region: string;
  public readonly accountId: string;
  public readonly privateSubnetIds: string[];
  public readonly publicSubnetIds: string[];
  public readonly secretsManagerSecretKey: kms.DataAwsKmsAlias;
  public readonly defaultSecurityGroups: vpc.DataAwsSecurityGroups;
  public readonly internalSecurityGroups: vpc.DataAwsSecurityGroups;

  constructor(scope: Construct, name: string, awsProvider?: AwsProvider) {
    super(scope, name);

    const vpcSSMParam = new ssm.DataAwsSsmParameter(this, `vpc_ssm_param`, {
      provider: awsProvider,
      name: '/Shared/Vpc',
    });

    this.vpc = new vpc.DataAwsVpc(this, `vpc`, {
      provider: awsProvider,
      filter: [
        {
          name: 'vpc-id',
          values: [vpcSSMParam.value],
        },
      ],
    });

    const privateString = new ssm.DataAwsSsmParameter(this, `private_subnets`, {
      provider: awsProvider,
      name: '/Shared/PrivateSubnets',
    });

    const privateSubnets = new vpc.DataAwsSubnetIds(
      this,
      `private_subnet_ids`,
      {
        provider: awsProvider,
        vpcId: this.vpc.id,
        filter: [
          {
            name: 'subnet-id',
            values: Fn.split(',', privateString.value),
          },
        ],
      }
    );

    this.privateSubnetIds = privateSubnets.ids;

    const publicString = new ssm.DataAwsSsmParameter(this, `public_subnets`, {
      provider: awsProvider,
      name: '/Shared/PublicSubnets',
    });

    const publicSubnets = new vpc.DataAwsSubnetIds(this, `public_subnet_ids`, {
      provider: awsProvider,
      vpcId: this.vpc.id,
      filter: [
        {
          name: 'subnet-id',
          values: Fn.split(',', publicString.value),
        },
      ],
    });

    this.publicSubnetIds = publicSubnets.ids;

    const identity = new datasources.DataAwsCallerIdentity(
      this,
      `current_identity`,
      {
        provider: awsProvider,
      }
    );
    this.accountId = identity.accountId;

    const region = new datasources.DataAwsRegion(this, 'current_region', {
      provider: awsProvider,
    });
    this.region = region.name;

    this.secretsManagerSecretKey = new kms.DataAwsKmsAlias(
      this,
      'secrets_manager_key',
      {
        provider: awsProvider,
        name: 'alias/aws/secretsmanager',
      }
    );

    this.defaultSecurityGroups = new vpc.DataAwsSecurityGroups(
      this,
      'default_security_groups',
      {
        provider: awsProvider,
        filter: [
          {
            name: 'group-name',
            values: ['default'],
          },
          {
            name: 'vpc-id',
            values: [this.vpc.id],
          },
        ],
      }
    );

    this.internalSecurityGroups = new vpc.DataAwsSecurityGroups(
      this,
      'internal_security_groups',
      {
        provider: awsProvider,
        filter: [
          {
            name: 'group-name',
            values: ['pocket-vpc-internal'],
          },
          {
            name: 'vpc-id',
            values: [this.vpc.id],
          },
        ],
      }
    );

  }
}
