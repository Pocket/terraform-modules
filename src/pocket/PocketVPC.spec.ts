import { Testing } from 'cdktf';
import { PocketVPC } from './PocketVPC';
import { VPC } from '@cdktf/provider-aws';
import 'cdktf/lib/testing/adapters/jest';

test('renders a VPC with minimal config', () => {
  const synthed = Testing.synthScope((stack) => {
    new PocketVPC(stack, 'testPocketVPC');
  });

  expect(synthed).toMatchSnapshot();
  expect(synthed).toHaveDataSource(VPC.DataAwsVpc);
});
