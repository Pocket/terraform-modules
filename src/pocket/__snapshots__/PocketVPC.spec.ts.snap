// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a VPC with minimal config 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"data\\": {
    \\"aws_ssm_parameter\\": {
      \\"testPocketVPC_vpc_ssm_param_DED34ABC\\": {
        \\"name\\": \\"/Shared/Vpc\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testPocketVPC/vpc_ssm_param\\",
            \\"uniqueId\\": \\"testPocketVPC_vpc_ssm_param_DED34ABC\\"
          }
        }
      },
      \\"testPocketVPC_private_subnets_3CC52CDC\\": {
        \\"name\\": \\"/Shared/PrivateSubnets\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testPocketVPC/private_subnets\\",
            \\"uniqueId\\": \\"testPocketVPC_private_subnets_3CC52CDC\\"
          }
        }
      },
      \\"testPocketVPC_public_subnets_21F01977\\": {
        \\"name\\": \\"/Shared/PublicSubnets\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testPocketVPC/public_subnets\\",
            \\"uniqueId\\": \\"testPocketVPC_public_subnets_21F01977\\"
          }
        }
      }
    },
    \\"aws_vpc\\": {
      \\"testPocketVPC_vpc_52CB41F5\\": {
        \\"filter\\": [
          {
            \\"name\\": \\"vpc-id\\",
            \\"values\\": [
              \\"\${data.aws_ssm_parameter.testPocketVPC_vpc_ssm_param_DED34ABC.value}\\"
            ]
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testPocketVPC/vpc\\",
            \\"uniqueId\\": \\"testPocketVPC_vpc_52CB41F5\\"
          }
        }
      }
    },
    \\"aws_subnet_ids\\": {
      \\"testPocketVPC_private_subnet_ids_61A307A3\\": {
        \\"vpc_id\\": \\"\${data.aws_vpc.testPocketVPC_vpc_52CB41F5.id}\\",
        \\"filter\\": [
          {
            \\"name\\": \\"subnet-id\\",
            \\"values\\": \\"\${split(\\\\\\",\\\\\\", data.aws_ssm_parameter.testPocketVPC_private_subnets_3CC52CDC.value)}\\"
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testPocketVPC/private_subnet_ids\\",
            \\"uniqueId\\": \\"testPocketVPC_private_subnet_ids_61A307A3\\"
          }
        }
      },
      \\"testPocketVPC_public_subnet_ids_5AC8BE89\\": {
        \\"vpc_id\\": \\"\${data.aws_vpc.testPocketVPC_vpc_52CB41F5.id}\\",
        \\"filter\\": [
          {
            \\"name\\": \\"subnet-id\\",
            \\"values\\": \\"\${split(\\\\\\",\\\\\\", data.aws_ssm_parameter.testPocketVPC_public_subnets_21F01977.value)}\\"
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testPocketVPC/public_subnet_ids\\",
            \\"uniqueId\\": \\"testPocketVPC_public_subnet_ids_5AC8BE89\\"
          }
        }
      }
    },
    \\"aws_caller_identity\\": {
      \\"testPocketVPC_current_identity_433574E3\\": {
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testPocketVPC/current_identity\\",
            \\"uniqueId\\": \\"testPocketVPC_current_identity_433574E3\\"
          }
        }
      }
    },
    \\"aws_region\\": {
      \\"testPocketVPC_current_region_1C1E28CD\\": {
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testPocketVPC/current_region\\",
            \\"uniqueId\\": \\"testPocketVPC_current_region_1C1E28CD\\"
          }
        }
      }
    },
    \\"aws_kms_alias\\": {
      \\"testPocketVPC_secrets_manager_key_3F5AECFC\\": {
        \\"name\\": \\"alias/aws/secretsmanager\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testPocketVPC/secrets_manager_key\\",
            \\"uniqueId\\": \\"testPocketVPC_secrets_manager_key_3F5AECFC\\"
          }
        }
      }
    },
    \\"aws_security_groups\\": {
      \\"testPocketVPC_default_security_groups_48E7EA82\\": {
        \\"filter\\": [
          {
            \\"name\\": \\"group-name\\",
            \\"values\\": [
              \\"default\\"
            ]
          },
          {
            \\"name\\": \\"vpc-id\\",
            \\"values\\": [
              \\"\${data.aws_vpc.testPocketVPC_vpc_52CB41F5.id}\\"
            ]
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testPocketVPC/default_security_groups\\",
            \\"uniqueId\\": \\"testPocketVPC_default_security_groups_48E7EA82\\"
          }
        }
      }
    }
  }
}"
`;
